### PR TITLE
feat: Add initial restructured text docstirng parsing

### PR DESCRIPTION
**Pull request by plannigan**
This implementation should be mostly complete.

Not implemented:
 - references: Not planned at this time
 - examples: I'm not aware of a standard format for this

Still needed:
 - more test:
   - Many of the error cases aren't currently covered
   - More full fixture test cases

To be cleaned up:
 - I'm not confident I got the types right for all of the data that comes in with the context
 - ParseContext works, but it is a bit awkward with the rest of the project using the dict

References: #67

### CODE REVIEW

**Comment by plannigan** [comment]
The PR still ended up being fairly big. Holding off on one of the main directives wouldn't have reduced the line count much. This seemed like a good stopping point to get something up for review. Subsequent PRs will be much smaller.

**Comment by pawamoy** [comment]
Hello! Sorry for the delay, I've quite busy :slightly_smiling_face: 

Thank you very much for this PR! I'm sure a lot of people will like it.
I'll start reviewing now.

**Comment by plannigan** [comment]
As a heads up, I tried using this implementation against a real project and it was not getting the `signature` of functions, so parameters were not displaying their `type` or `kind` correctly. I followed the same pattern used for the google format for parsing in the unit tests. But I might have missed something related to the context dictionary. So keep an eye of for that even though all of the unit tests currently pass.

**Review by pawamoy — CHANGES_REQUESTED** [review]
I think I'm a bit too tired to read all that code closely :sweat_smile: 

But with a quick look I can tell the code is clean, you added comments to explain things, so LGTM.

Just a nitpick about the helpers in the test file.

About the signature of functions not being passed, maybe you could try to write a test that is actually loading an object, to see how it behaves from A to Z. I started thinking about the code structure, hopefully I'll find some time to improve it so it will be more straight-forward and easier to reason about.

**Comment by pawamoy** [comment]
Oh, it would also be really great if you could update the README :slightly_smiling_face:

File: tests/test_parsers/test_docstrings/test_restructured_text.py
Comment (plannigan) [reply tests/test_parsers/test_docstrings/test_restructured_text.py:699]: Sure. I generally do it with the helper functions at the end, but it doesn't matter much. Another option would be to implement `__eq__()` on the section detail objects, This would allow for `assert sections[1].value == [EXPECTED_PARAM]`. I opted not to do that to keep my PR more self-contained without making changes to the other structures.

**Comment by plannigan** [comment]
I was able to track down there error in my code related to the context. (A helper function in the google test suite [passes the signature and type in as part of the context](https://github.com/pawamoy/pytkdocs/blob/1a80c5b13eab837469e7a189f77212c6cff230c2/tests/test_parsers/test_docstrings/test_google.py#L19). However, the main source code only ever passes in `obj` and optionally `attributes`. I altered the helper function to be similar to what the main source does.)

I also added more test cases complete coverage.

**Comment by plannigan** [comment]
Wanted to check back in. Is there anything else you would like me to add to this PR before it can be merged?

**Comment by pawamoy** [comment]
Hello! I've been lacking time and energy... but I'll try to give a final review tomorrow :slightly_smiling_face:

**Pull request merged**
